### ISSUE TITLE
Request caching might be a bottleneck for concurrent cdk

### ISSUE THREAD

**Issue by girarda** [airbytehq/airbyte#31585]
## What
The current implementation of request caching uses the `requests_cache` library with a sqlite backing.
Writes to the cache appear to be synchronous and blocking since running the stripe connector with >= 4 workers leads to writes to the cache being skipped.
Sample warning:
`Database is locked in thread 6153383936; retrying (1/3)`

My current understanding is this can lead to sync failures if all 3 tries fail.
[It's also possible the cache just bails without writing the entry](https://github.com/requests-cache/requests-cache/blob/b75a9f6c61bcd45a3bf55c940d61580bc025314a/requests_cache/backends/sqlite.py#L324-L341).

## Acceptance criteria
- [ ] We understand the impact of failing to access the db 3 times
- [ ] We verified the concurrent cdk is not bottlenecked by the request caching
 - [ ] If it is, a fix is either implemented or described in a follow up issue

**Comment by girarda** [comment]
grooming notes:
- requests_cache with sqlite has a WAL feature
- we should also verify the performance cost of removing caching for source stripe (we should check full refresh + incremental)
- alternative: look into other backends available. we want one that is threadsafe

**Comment by girarda** [comment]
confirming the syncs will fail if a thread can't connector to the db because it is locked. 

sample error message for posterity:
```
{"type": "LOG", "log": {"level": "ERROR", "message": "Encountered an exception while reading stream application_fees_refunds\nTraceback (most recent call last):\n  File \"/Users/alex/code/airbyte/airbyte-cdk/python/airbyte_cdk/sources/abstract_source.py\", line 116, in read\n    stream_is_available, reason = stream_instance.check_availability(logger, self)\n  File \"/Users/alex/code/airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/adapters.py\", line 178, in check_availability\n    availability = self._abstract_stream.check_availability()\n  File \"/Users/alex/code/airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/thread_based_concurrent_stream.py\", line 163, in check_availability\n    return self._availability_strategy.check_availability(self._logger)\n  File \"/Users/alex/code/airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/adapters.py\", line 331, in check_availability\n    raise e\n  File \"/Users/alex/code/airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/adapters.py\", line 321, in check_availability\n    available, message = self._stream.check_availability(logger, self._source)\n  File \"/Users/alex/code/airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/core.py\", line 211, in check_availability\n    return self.availability_strategy.check_availability(self, logger, source)\n  File \"/Users/alex/code/airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/http/availability_strategy.py\", line 36, in check_availability\n    stream_slice = get_first_stream_slice(stream)\n  File \"/Users/alex/code/airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/utils/stream_helper.py\", line 26, in get_first_stream_slice\n    return next(slices)\n  File \"/Users/alex/code/airbyte/airbyte-integrations/connectors/source-stripe/source_stripe/streams.py\", line 854, in stream_slices\n    yield from self.parent_stream.stream_slices(sync_mode, cursor_field=cursor_field, stream_state=stream_state)\n  File \"/Users/alex/code/airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py\", line 516, in stream_slices\n    for record in parent_records:\n  File \"/Users/alex/code/airbyte/airbyte-integrations/connectors/source-stripe/source_stripe/streams.py\", line 470, in read_records\n    yield from self.parent_stream.read_records(sync_mode, cursor_field, stream_slice, stream_state)\n  File \"/Users/alex/code/airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py\", line 448, in read_records\n    yield from self._read_pages(\n  File \"/Users/alex/code/airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py\", line 464, in _read_pages\n    request, response = self._fetch_next_page(stream_slice, stream_state, next_page_token)\n  File \"/Users/alex/code/airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py\", line 490, in _fetch_next_page\n    response = self._send_request(request, request_kwargs)\n  File \"/Users/alex/code/airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py\", line 388, in _send_request\n    return backoff_handler(user_backoff_handler)(request, request_kwargs)\n  File \"/Users/alex/code/airbyte/airbyte-cdk/python/.venv/lib/python3.10/site-packages/backoff/_sync.py\", line 105, in retry\n    ret = target(*args, **kwargs)\n  File \"/Users/alex/code/airbyte/airbyte-cdk/python/.venv/lib/python3.10/site-packages/backoff/_sync.py\", line 105, in retry\n    ret = target(*args, **kwargs)\n  File \"/Users/alex/code/airbyte/airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py\", line 332, in _send\n    response: requests.Response = self._session.send(request, **request_kwargs)\n  File \"/Users/alex/code/airbyte/airbyte-cdk/python/.venv/lib/python3.10/site-packages/requests_cache/session.py\", line 193, in send\n    cached_response = self.cache.get_response(actions.cache_key)\n  File \"/Users/alex/code/airbyte/airbyte-cdk/python/.venv/lib/python3.10/site-packages/requests_cache/backends/base.py\", line 70, in get_response\n    response = self.responses[self.redirects[key]]\n  File \"/Users/alex/code/airbyte/airbyte-cdk/python/.venv/lib/python3.10/site-packages/requests_cache/backends/sqlite.py\", line 294, in __getitem__\n    cur = con.execute(f'SELECT value FROM {self.table_name} WHERE key=?', (key,))\nsqlite3.OperationalError: database is locked"}}
```

**Comment by girarda** [comment]
I ran a couple of tests with concurrent stripe locally.
<img width="634" alt="Screenshot 2023-10-19 at 2 50 13 PM" src="https://github.com/airbytehq/airbyte/assets/1451943/20922023-490d-48ea-a97d-240b9e78245e">

the gist of the results are that
- request caching help save about half the time (with the specific configuration I used)
- the fast_save and wal flag on the SQLiteCache seem to have a small impact on performance, but the difference might be smaller than the variance between runs. This requires more investigation

None of the runs triggered a `Database is locked` warning so I can't say that either setting help with the issue.
